✨ (kustomize/v2): Add NamespaceTransformer to support multi-namespace…

AlirezaPourchali opened the pull request:
## Description

This PR fixes namespace conflicts when using namespace-scoped RBAC markers in multi-namespace scenarios.

### Problem
When using RBAC markers with explicit namespaces:
```go
// +kubebuilder:rbac:groups=apps,namespace=infrastructure,resources=deployments,verbs=get
// +kubebuilder:rbac:groups="",namespace=users,resources=secrets,verbs=get
```

The `controller-gen` correctly generates separate Role resources for each namespace. However, during `make deploy`, Kustomize's global `namespace:` field in `config/default/kustomization.yaml` overrides ALL namespaces, causing both Roles to end up in the same namespace with identical names → ID conflict error.

### Solution
Replaces the hardcoded `namespace:` field with a `NamespaceTransformer` using `unsetOnly: true`. This approach:
- ✅ Preserves explicit namespaces from RBAC markers
- ✅ Adds default namespace to resources that don't specify one
- ✅ Maintains backward compatibility (users can revert by uncommenting the old `namespace:` field)
- ✅ Follows Kustomize best practices per [NamespaceTransformer docs](https://kubectl.docs.kubernetes.io/references/kustomize/builtins/#_namespacetransformer_)

### Changes
1. **Created**: `pkg/plugins/common/kustomize/v2/scaffolds/internal/templates/config/kdefault/namespace_transformer.go`
   - Scaffolds `namespace-transformer.yaml` with `unsetOnly: true`
   
2. **Modified**: `pkg/plugins/common/kustomize/v2/scaffolds/internal/templates/config/kdefault/kustomization.go`
   - Commented out hardcoded `namespace:` field
   - Added `transformers:` section referencing `namespace-transformer.yaml`
   - Included documentation comments explaining the change and fallback option

3. **Modified**: `pkg/plugins/common/kustomize/v2/scaffolds/init.go`
   - Registered `NamespaceTransformer` in templates slice

### ⚠️ Known Side Effect: Helm Charts

The Helm chart generation was affected by this change. Some generated Helm templates changed from:
```yaml
namespace: {{ .Release.Namespace }}
```
to:
```yaml  
namespace: system
```

**Root cause**: The `metrics_service.yaml` template has a hardcoded `namespace: system` field. With `unsetOnly: true`, the NamespaceTransformer preserves this existing namespace. The Helm v2-alpha plugin's `substituteNamespace()` function in `helm_templater.go` expects the format `<projectname>-system` (e.g., `project-v4-system`) but encounters just `system`, so it doesn't convert it to `{{ .Release.Namespace }}`.

**Question for maintainers**: Should the Helm v2-alpha plugin be updated to also replace the `namespace: system` pattern? I can make that change if needed, or if there's a preferred alternative solution. The v1-alpha plugin handles this via `strings.ReplaceAll(contentStr, "namespace: system", "namespace: {{ .Release.Namespace }}")` in `edit.go:442`.

### Testing
- ✅ `make generate` - All testdata and docs regenerated successfully
- ✅ `make lint-fix` - No linting issues
- ✅ `make test-unit` - All unit tests pass
- ✅ Manual verification of generated `namespace-transformer.yaml` in test projects

### Backward Compatibility
Existing projects are unaffected. New projects created with `kubebuilder init` will use the NamespaceTransformer approach. Users who prefer the old behavior can:
1. Uncomment `namespace: {{ .ProjectName }}-system` in `config/default/kustomization.yaml`
2. Remove the `transformers:` section
3. Delete `namespace-transformer.yaml`

Fixes #5148

camilamacedo86 commented on docs/book/src/cronjob-tutorial/testdata/project/dist/chart/templates/prometheus/controller-manager-metrics-monitor.yaml:
```diff
@@ -7,7 +7,7 @@ metadata:
         control-plane: controller-manager
     name: project-controller-manager-metrics-monitor
-    namespace: {{ .Release.Namespace }}
+    namespace: system
```
We cannot do those changes in the Helm Charts.

AlirezaPourchali replied:
**To be clear: this Helm side effect is a bug that needs to be fixed.** The changes in the Helm charts are unintended consequences of the NamespaceTransformer approach. The core Kustomize fix is correct, but it exposed an existing inconsistency in how we handle the metrics service namespace.

## Why This Worked Before But Breaks Now

**Before (with hardcoded `namespace:` field):**
1. `metrics_service.yaml` generated with `namespace: system`
2. Kustomize's hardcoded `namespace: project-v4-system` **forcefully overrode** ALL namespaces
3. Final output: `namespace: project-v4-system` (not `system`)
4. Helm plugin saw `project-v4-system`, matched the pattern, replaced with `{{ .Release.Namespace }}` ✅

**Now (with NamespaceTransformer + `unsetOnly: true`):**
1. `metrics_service.yaml` generated with `namespace: system`
2. NamespaceTransformer sees namespace **already set**, so **preserves it** (because `unsetOnly: true`)
3. Final output: `namespace: system` (unchanged)
4. Helm plugin sees `system`, doesn't match `project-v4-system` pattern, no replacement ❌

**The aggressive namespace override was masking an existing inconsistency.** The `metrics_service.go` template should have been using the same pattern as other resources all along, but nobody noticed because the hardcoded `namespace:` field was forcing everything to the correct value.

Now that we use `unsetOnly: true` (which correctly preserves explicit namespaces for multi-namespace RBAC), the inconsistency is exposed.

## The Root Cause

The issue is in how `metrics_service.go` template is defined. Currently:

**File: `pkg/plugins/common/kustomize/v2/scaffolds/internal/templates/config/kdefault/metrics_service.go`** (line 52)

```yaml
metadata:
  name: controller-manager-metrics-service
  namespace: system    # ← Hardcoded "system"
```

The Helm v2-alpha plugin's `substituteNamespace()` function expects the pattern `<projectname>-system` (e.g., `project-v4-system`):

**File: `pkg/plugins/optional/helm/v2alpha/scaffolds/internal/kustomize/helm_templater.go`** (line 114)

```go
func (t *HelmTemplater) substituteNamespace(yamlContent string, resource *unstructured.Unstructured) string {
	hardcodedNamespace := t.projectName + "-system"  // ← Expects "project-v4-system"
	namespaceTemplate := "{{ .Release.Namespace }}"
	
	yamlContent = strings.ReplaceAll(yamlContent, hardcodedNamespace, namespaceTemplate)
	return yamlContent
}
```

When it encounters just `system`, it doesn't match the expected pattern, so it doesn't get replaced with `{{ .Release.Namespace }}`.

## Two Options to Fix

### Option 1: Fix metrics_service.go Template (Recommended)

Change the hardcoded `system` to use the project name pattern:

```go
const metricsServiceTemplate = `apiVersion: v1
kind: Service
metadata:
  labels:
    control-plane: controller-manager
    app.kubernetes.io/name: {{ .ProjectName }}
    app.kubernetes.io/managed-by: kustomize
  name: controller-manager-metrics-service
  namespace: {{ .ProjectName }}-system    # ← Use project name pattern
spec:
  ...
```

This way:
- Kustomize with NamespaceTransformer will preserve `{{ .ProjectName }}-system` (it's a template variable, not a literal namespace)
- Helm v2-alpha plugin will match the pattern and convert to `{{ .Release.Namespace }}`

### Option 2: Update Helm v2-alpha Plugin

Add a fallback in `substituteNamespace()` to also handle plain `system`:

```go
func (t *HelmTemplater) substituteNamespace(yamlContent string, resource *unstructured.Unstructured) string {
	hardcodedNamespace := t.projectName + "-system"
	namespaceTemplate := "{{ .Release.Namespace }}"
	
	yamlContent = strings.ReplaceAll(yamlContent, hardcodedNamespace, namespaceTemplate)
	
	// Fallback: also replace plain "system" (for metrics service)
	yamlContent = strings.ReplaceAll(yamlContent, "namespace: system", "namespace: "+namespaceTemplate)
	
	return yamlContent
}
```

## My Recommendation

I think **Option 1 is cleaner** - the `metrics_service.go` template should follow the same pattern as other resources. This would make the Helm plugin's logic simpler and more consistent.

**This needs to be fixed before merging this PR.** The core NamespaceTransformer changes are sound, but we can't introduce this Helm regression.

Should I:
1. Update `metrics_service.go` to use `{{ .ProjectName }}-system` pattern and regenerate testdata? (My preference)
2. Update the Helm v2-alpha plugin with the fallback for plain `system`?
3. Revert to the hardcoded `namespace:` field approach until we can coordinate both fixes?

I'm ready to implement whichever solution you prefer. Let me know and I'll get it done! 🙏

camilamacedo86 reviewed (changes requested):
Hi @AlirezaPourchali 👋

Thank you for raising this and for the work on it — much appreciated.

I think the first step should be opening a PR against controller-tools.
See this comment for context:
https://github.com/kubernetes-sigs/kubebuilder/issues/5148#issuecomment-3536838419

Once we have the fix in controller-runtime, we can then evaluate what is needed in kubebuilder. Having the fix in place will allow us to properly test and make sure the solution really works.

With the controller-runtime fix, I don’t think we need all of these changes in kubebuilder.
It seems we would only need one additional file:
config/default/namespace-transformer.yaml
(as shown in my example comment).

Also, instead of adding this to the default scaffold, it might be better to document it. This is probably not needed for ~90% of users, and adding a new default file may not help most people.

Possible alternatives:
- Add documentation in the FAQ: https://book.kubebuilder.io/faq
- Or create a small reference/tutorial explaining this use case

We could include a concrete example (similar to the CronJob references), with:
- a small mock project
- e2e tests
- and Helm validation
to ensure the approach works end-to-end.

What do you think? Does this approach make sense?

Thanks again for the contribution 🙏

camilamacedo86 commented:
Hi @AlirezaPourchali,

This is really helpful—thank you so much for the thorough explanations and for sharing your tests and research. I’d like a bit of time to review everything properly, and I’ll reply back here once I’ve gone through it.

AlirezaPourchali commented:
Hi @camilamacedo86, Could you checkout my changes.

AlirezaPourchali committed:
feat(kustomize/v2): scaffold NamespaceTransformer for multi-namespace RBAC

Scaffold a commented-out NamespaceTransformer configuration as an opt-in
alternative to the "namespace:" field in config/default/kustomization.yaml.

When using namespace-scoped RBAC markers (e.g.
//+kubebuilder:rbac:namespace=infrastructure), the global "namespace:" field
overrides all namespaces, causing ID conflict errors. The NamespaceTransformer
with "unsetOnly: true" preserves explicit namespaces while assigning the
default namespace to resources without one.

Changes:
- Add namespace_transformer.go template scaffolding namespace_transformer.yaml
- Add commented-out [NAMESPACE-TRANSFORMER] section in kustomization.yaml with
  inline documentation explaining the ID conflict issue and opt-in steps
- Add NOTE comment on the "namespace:" field warning about the override behavior
- Register NamespaceTransformer in init.go

Default behavior is completely unchanged. Users who need multi-namespace RBAC
can follow the inline instructions to switch from the "namespace:" field to
the NamespaceTransformer.
